ingestion/src/metadata/ingestion/connections/test_connections.py

```diff
@@ -89,35 +209,135 @@ def test_connection_steps(steps: List[TestConnectionStep]) -> TestConnectionResu
+    logger.info(test_connection_result)
 
+    if test_connection_result.failed:
+        raise SourceConnectionException(
+            f"Some steps failed when testing the connection: [{test_connection_result}]"
+        )
 
-def test_connection_engine(connection: Engine, steps=None) -> TestConnectionResult:
-    try:
-        with connection.connect() as conn:
-            conn.execute(ConnTestFn())
-            if steps:
-                return test_connection_steps(steps)
-    except SourceConnectionException as exc:
-        raise exc
-    except OperationalError as err:
-        msg = f"Connection error for {connection}: {err}. Check the connection details."
-        raise SourceConnectionException(msg) from err
-    except Exception as exc:
-        msg = f"Unknown error connecting with {connection}: {exc}."
-        raise SourceConnectionException(msg) from exc
-
-    return None
+
+def test_connection_steps(
+    metadata: OpenMetadata,
+    service_fqn: str,
+    test_fn: dict,
+    automation_workflow: Optional[AutomationWorkflow] = None,
+    timeout_seconds: int = 3 * 60,
+) -> None:
+    """
+    Test the connection steps with a given timeout
+
+    Test that we can connect to the source using the given connection
+    and extract the metadata
+    :return: None or raise an exception if we cannot connect
+    """
+
+    test_connection_definition: TestConnectionDefinition = metadata.get_by_name(
+        entity=TestConnectionDefinition,
+        fqn=service_fqn,
+    )
+
+    steps = [
+        TestConnectionStep(
+            name=step.name,
+            description=step.description,
+            mandatory=step.mandatory,
+            function=test_fn[step.name],
+        )
+        for step in test_connection_definition.steps
+    ]
+
+    return timeout(timeout_seconds)(_test_connection_steps)(
+        metadata, steps, automation_workflow
+    )
+
+
+def test_connection_engine_step(connection: Engine) -> None:
+    """
+    Generic step to validate the connection against a db
+    """
+    with connection.connect() as conn:
+        conn.execute(ConnTestFn())
 
 
 def test_connection_db_common(
-    connection: Engine, steps=None, timeout_seconds: int = 120
+    metadata: OpenMetadata,
+    engine: Engine,
+    service_connection,
+    automation_workflow: Optional[AutomationWorkflow] = None,
+    queries: dict = {},
+    timeout_seconds: int = 3 * 60,
 ) -> TestConnectionResult:
+
     """
-    Default implementation is the engine to test.
+    Test connection
+    """
+    inspector = inspect(engine)
 
-    Test that we can connect to the source using the given engine
-    :param connection: Engine to test
-    :return: None or raise an exception if we cannot connect
+    test_fn = {
+        "CheckAccess": partial(test_connection_engine_step, engine),
+        "GetSchemas": inspector.get_schema_names,
+        "GetTables": inspector.get_table_names,
+        "GetViews": inspector.get_view_names,
+    }
```

Comment:
How can we overwrite any of these values or in case we want to add more test functions?

Reply:
Umm the simple answer would be that we can not, most of the databases supports all these operation if you have to add or subtract anything then you can checkout how it is done for bigquery where we are addition step to test tags